README.md
# ravioly :spaghetti:

[![](https://img.shields.io/badge/linkedin-connect-9cf?&logo=linkedin)](https://www.linkedin.com/in/yrieix.leprince)
![](https://img.shields.io/github/license/yleprince/ravioly.svg?color=9cf)
![](https://img.shields.io/github/languages/top/yleprince/ravioly.svg)

![](https://github.com/yleprince/ravioly/workflows/Python%20CI/badge.svg)
![](https://img.shields.io/github/last-commit/yleprince/ravioly.svg)
![](https://img.shields.io/github/commit-activity/y/yleprince/ravioly)

![](https://img.shields.io/github/issues-pr-closed-raw/yleprince/ravioly.svg)
![](https://img.shields.io/github/issues-closed-raw/yleprince/ravioly.svg)


## Context :car: :cityscape: :us:
The goal is to have a production ready code to analyse New York taxi data.

Ravioly is a way to encapsulate the raw data contained in NYC taxi csv files. As mentionned in the documentation (https://yleprince.github.io/ravioly/) it is built on top of `pandas.DataFrame` and provide specific processing functionalities and dedicated methods.

## Install and Use: :seedling:

:bulb: Python `^3.7` is required.

Install using `pip`:

```sh
pip install git+https://github.com/yleprince/ravioly.git
```

In your python code:

```python
>>> from ravioly.datastructure import Ravioly

>>> df = Ravioly('../data/nyc_data.csv', nrows=1000)
>>> df.km_by_dow()

day_of_week
0    480.647876
1    466.137703
2    553.287868
3    427.187865
4    465.982398
5    489.352866
6    557.113716
Name: km_by_dow, dtype: float64
```

## Tools used: :gear:
* **`Dev`**: :computer:
  * [pandas](https://pandas.pydata.org/): the code is a subclass of pandas.DataFrame data structure.
  * [poetry](https://python-poetry.org/): package management is made with poetry
  
* **`Lint`**: :triangular_ruler: 
  * [isort](https://pypi.org/project/isort/): to sort the imports
  * [mypy](http://mypy-lang.org/): to check the use of types within the code
  * [black](https://pypi.org/project/black/): to format automatically the code with the pep8 requirements
  * [flake8](https://gitlab.com/PyCQA/flake8): to check code syntax

* **`Tests`**: :teacher:
  * [pytest](https://docs.pytest.org/en/latest/): to unit test the code
  * [pytest-cov](https://pypi.org/project/pytest-cov/): to check percentage of code covered by the tests

* **`CI`**: :robot:
  * [github actions](https://github.com/features/actions): to maintain code consistency over commits.

* **`Documentation`**: :books: 
  * [sphinx](https://www.sphinx-doc.org/en/master/): automatically generate the documentation from source code.
  * [read the doc theme](https://sphinx-rtd-theme.readthedocs.io/): theme for the documentation
  * [github pages](https://pages.github.com/): to serve the documentation: https://yleprince.github.io/ravioly/

* **`Bonus`**: :gift: 
  * [pre-commit](https://pre-commit.com/): pre-commit allows to run lint and tests workflow automatically at every step of the project
  * [pre-push](https://www.git-scm.com/docs/githooks#_pre_push): pre-push allows to update the documentation every time the code is pushed on the github.
 


CHANGELOG.md
# # CHANGELOG

## [Unreleased]

## v0.1.0

* Initial release
* distance computation, with Haversine formula


tests/test_ravioli.py
import os
import unittest
from typing import List

import pandas as pd

from ravioly.datastructure import Ravioly


class TestInitravioly(unittest.TestCase):
    """Checks that the ravioly class succeeds during init"""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_init(self):
        df: ravioly = Ravioly(self.filepath)
        self.assertIsInstance(df, pd.DataFrame)
        self.assertIsInstance(df, Ravioly)
        self.assertEqual(df.shape, (40, 12))

    def test_init_args(self):
        columns: List[str] = [
            "pickup_datetime",
            "dropoff_datetime",
            "pickup_longitude",
            "pickup_latitude",
            "dropoff_longitude",
            "dropoff_latitude",
            "trip_duration",
        ]
        df: Ravioly = Ravioly(self.filepath, nrows=26, usecols=columns)
        self.assertIsInstance(df, pd.DataFrame)
        self.assertIsInstance(df, Ravioly)
        self.assertEqual(df.shape, (26, 8))
        self.assertTrue("distance" in df.columns)


class TestDistanceComputations(unittest.TestCase):
    """Checks that the Ravioly class succeeds during init to compute the distance
     between the pickup and the drop off coordinates."""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_distance_values(self):
        df: Ravioly = Ravioly(self.filepath)
        expected: List[float] = [
            1.498521,
            1.805507,
            6.385098,
            1.485498,
            1.188588,
            1.098942,
            1.326279,
            5.714981,
            1.310353,
            5.121162,
            3.806139,
            3.773096,
            1.859483,
            0.991685,
            6.382836,
            0.656578,
            3.428086,
            2.538672,
            4.605201,
            1.303271,
            2.505926,
            1.724550,
            2.067085,
            4.874792,
            20.602575,
            4.559525,
            6.056109,
            3.738742,
            2.524849,
            9.939443,
            4.564593,
            5.423052,
            4.499284,
            10.238385,
            1.433100,
            1.492237,
            1.136392,
            0.959435,
            3.312410,
            2.587523,
        ]
        for value, target in zip(df.distance.values, expected):
            self.assertAlmostEqual(value, target, delta=0.0001)

    def test_haversine_distance(self):
        self.assertEqual(Ravioly._haversine_distance(self, (0, 0), (0, 0)), 0.0)
        self.assertEqual(
            Ravioly._haversine_distance(self, (48.87, 2.33), (51.53, -0.24)),
            347.72272585658754,
        )


tests/test_ravioly.py
import os
import unittest
from typing import List

import pandas as pd
from ravioly.datastructure import Ravioly


class TestInitRavioly(unittest.TestCase):
    """Tests that the Ravioly class succeeds during init"""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_init(self):
        df: Ravioly = Ravioly(self.filepath)
        self.assertIsInstance(df, pd.DataFrame)
        self.assertIsInstance(df, Ravioly)
        self.assertEqual(df.shape, (40, 13))

    def test_init_args(self):
        columns: List[str] = [
            "pickup_datetime",
            "dropoff_datetime",
            "pickup_longitude",
            "pickup_latitude",
            "dropoff_longitude",
            "dropoff_latitude",
            "trip_duration",
        ]
        df: Ravioly = Ravioly(self.filepath, nrows=26, usecols=columns)
        self.assertIsInstance(df, pd.DataFrame)
        self.assertIsInstance(df, Ravioly)
        self.assertEqual(df.shape, (26, 9))
        self.assertTrue("distance" in df.columns)
        self.assertTrue("avg_speed" in df.columns)


class TestDistanceComputations(unittest.TestCase):
    """Tests that the Ravioly class succeeds during init to compute the distance
     between the pickup and the drop off coordinates."""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_distance_values(self):
        df: Ravioly = Ravioly(self.filepath)
        expected: List[float] = [
            1.498521,
            1.805507,
            6.385098,
            1.485498,
            1.188588,
            1.098942,
            1.326279,
            5.714981,
            1.310353,
            5.121162,
            3.806139,
            3.773096,
            1.859483,
            0.991685,
            6.382836,
            0.656578,
            3.428086,
            2.538672,
            4.605201,
            1.303271,
            2.505926,
            1.724550,
            2.067085,
            4.874792,
            20.602575,
            4.559525,
            6.056109,
            3.738742,
            2.524849,
            9.939443,
            4.564593,
            5.423052,
            4.499284,
            10.238385,
            1.433100,
            1.492237,
            1.136392,
            0.959435,
            3.312410,
            2.587523,
        ]
        for value, target in zip(df.distance.values, expected):
            self.assertAlmostEqual(value, target, delta=0.0001)

    def test_haversine_distance(self):
        self.assertEqual(Ravioly._haversine_distance(self, (0, 0), (0, 0)), 0.0)

        # d(Paris - London) is 347 km
        self.assertEqual(
            Ravioly._haversine_distance(self, (48.87, 2.33), (51.53, -0.24)),
            347.72272585658754,
        )


class TestAverageSpeedComputations(unittest.TestCase):
    """Tests that the Ravioly class succeeds during init to compute average
     speeds between the pickup and the drop off coordinates."""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_avg_speed_values(self):
        df: Ravioly = Ravioly(self.filepath)
        expected: List[float] = [
            11.856428,
            9.803659,
            10.822201,
            12.465721,
            9.836594,
            8.930458,
            14.001768,
            13.264945,
            18.499105,
            15.049944,
            10.755182,
            12.041796,
            6.009101,
            13.731021,
            16.250501,
            11.202279,
            5.328631,
            12.502351,
            12.588249,
            18.692336,
            18.562416,
            9.522056,
            17.592212,
            15.089641,
            29.846789,
            12.793679,
            19.293800,
            19.394051,
            10.189974,
            15.350491,
            11.110572,
            18.628805,
            15.848751,
            16.244242,
            7.297254,
            10.213026,
            7.397850,
            12.469197,
            10.029165,
            6.116273,
        ]
        for value, target in zip(df.avg_speed.values, expected):
            self.assertAlmostEqual(value, target, delta=0.0001)


class TestTripByDayOfWeek(unittest.TestCase):
    """Tests the method trip_by_dow that counts the number of trips by weekday."""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_trip_by_dow(self):
        df: Ravioly = Ravioly(self.filepath)
        expected: List[int] = [6, 6, 3, 3, 8, 8, 6]
        for value, target in zip(df.trip_by_dow().values, expected):
            self.assertEqual(value, target)
        self.assertEqual(df.trip_by_dow().name, "trip_by_dow")


class TestTripByIHours(unittest.TestCase):
    """Tests the method trip_by_ih that counts the number of trips every `i` hours."""

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_trip_by_ih(self):
        df: Ravioly = Ravioly(self.filepath)

        expected_2h: List[int] = [2, 2, 2, 4, 4, 6, 5, 3, 2, 4, 6]
        for value, target in zip(df.trip_by_ih(2).values, expected_2h):
            self.assertEqual(value, target)
        self.assertEqual(df.trip_by_ih(2).name, "trip_by_2h")

        expected_23h: List[int] = [37, 3]
        for value, target in zip(df.trip_by_ih(23).values, expected_23h):
            self.assertEqual(value, target)
        self.assertEqual(df.trip_by_ih(23).name, "trip_by_23h")

        expected_24h: List[int] = [40]
        for value, target in zip(df.trip_by_ih(24).values, expected_24h):
            self.assertEqual(value, target)
        self.assertEqual(df.trip_by_ih(24).name, "trip_by_24h")

    def test_limits_trip_by_ih(self):
        df: Ravioly = Ravioly(self.filepath)
        with self.assertRaises(ValueError) as context:
            df.trip_by_ih(-1)
        self.assertTrue(
            "step should int or float, and within ]0, 24]. -1 given."
            in str(context.exception)
        )

        with self.assertRaises(ValueError) as context:
            df.trip_by_ih(25)
        self.assertTrue(
            "step should int or float, and within ]0, 24]. 25 given."
            in str(context.exception)
        )

        with self.assertRaises(ValueError) as context:
            df.trip_by_ih("toto")
        self.assertTrue(
            "step should int or float, and within ]0, 24]. toto given."
            in str(context.exception)
        )

    def test_trip_by_4h(self):
        df: Ravioly = Ravioly(self.filepath)
        expected: List[int] = [4, 2, 8, 11, 5, 10]
        for value, target in zip(df.trip_by_4h().values, expected):
            self.assertEqual(value, target)
        self.assertEqual(df.trip_by_4h().name, "trip_by_4h")


class TestKilometerByDayOfWeek(unittest.TestCase):
    """
    Tests the method km_by_dow that sums the number of kilometers traveled
    by day of week.
    """

    def setUp(self):
        self.path = os.path.dirname(__file__)
        self.filepath = os.path.join(self.path, "samples.csv")

    def test_km_by_dow(self):
        df: Ravioly = Ravioly(self.filepath)
        expected: List[float] = [
            16.45197715,
            22.97903081,
            12.41662616,
            8.61926878,
            38.01072274,
            27.85664986,
            24.18571021,
        ]

        for value, target in zip(df.km_by_dow().values, expected):
            self.assertAlmostEqual(value, target, delta=0.0001)


ravioly/datastructure.py
from math import atan2, cos, radians, sin, sqrt
from typing import Tuple

from pandas import DataFrame, Series, read_csv, to_datetime


class Ravioly(DataFrame):
    """
    **Ravioly is a pandas.DataFrame subclass dedicated to New York taxi dataset.**


    Automated processings
     * `pickup` and `dropoff` datetimes are converted in datetime type, and localized to NYC timezone.
     * `distance` and `avg_speed` are automatically generated from raw data.

    :Init:
     Ravioly init overloads `pd.read_csv`, so it has a mendatory parameter `filepath` that leads to the New York city csv file.

    :Example:

    >>> from ravioly.datastructure import Ravioly
    >>> Ravioly('../data/train.csv', nrows=2)
               id  vendor_id           pickup_datetime          dropoff_datetime  passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude store_and_fwd_flag  trip_duration  distance  avg_speed
    0  id2875421          2 2016-03-14 17:24:55-04:00 2016-03-14 17:32:30-04:00                 1        -73.982155        40.767937         -73.964630         40.765602                  N            455  1.498521  11.856428
    1  id2377394          1 2016-06-12 00:43:35-04:00 2016-06-12 00:54:38-04:00                 1        -73.980415        40.738564         -73.999481         40.731152                  N            663  1.805507   9.803659
    """

    def __init__(self, filepath: str, *args, **kwargs):
        super().__init__(read_csv(filepath, *args, **kwargs))
        self.pickup_datetime: Series = to_datetime(self.pickup_datetime).dt.tz_localize(
            "US/Eastern"
        )
        self.dropoff_datetime: Series = to_datetime(
            self.dropoff_datetime
        ).dt.tz_localize("US/Eastern")
        self["distance"] = None
        self["avg_speed"] = None
        self.distance: Series = self.apply(self._distance, axis="columns")
        self.avg_speed: Series = self.distance / (self.trip_duration / 3600)

    def _haversine_distance(
        self, coords0: Tuple[float, float], coords1: Tuple[float, float]
    ) -> float:
        """
        Implementation of the Haversine formula:
        https://en.wikipedia.org/wiki/Haversine_formula

        :param coords0: first GPS coordinates (latitude, longitude)
        :param coords1: second GPS coordinates (latitude, longitude)
        :return: haversine distance in KM between the two GPS coordinates.

        >>> _haversine_distance((0, 0), (0, 0))
        0.0
        >>> _haversine_distance((48.87, 2.33), (51.53, -0.24))
        347.72272585658754
        """
        radius: int = 6371  # Earth avg radius

        lat0: float = coords0[0]
        lon0: float = coords0[1]
        lat1: float = coords1[0]
        lon1: float = coords1[1]

        phi0: float = radians(lat0)
        phi1: float = radians(lat1)
        dphi: float = radians(lat1 - lat0)
        dlambda: float = radians(lon1 - lon0)

        a: float = sin(dphi / 2) ** 2 + cos(phi0) * cos(phi1) * sin(dlambda / 2) ** 2
        distance: float = 2 * radius * atan2(sqrt(a), sqrt(1 - a))
        return distance

    def _distance(self, row: Series) -> float:
        """
        Compute distance for a single row (ie. taxi trip) using the pickup and
        dropoff coordinates.

        :param row: row of the Ravioly, must contain `pickup_latitude` and `pickup_longitude` columns
        :type row: pandas.Series
        :return: distance in kilometers
        """
        coords0: Tuple[float, float] = (row.pickup_latitude, row.pickup_longitude)
        coords1: Tuple[float, float] = (row.dropoff_latitude, row.dropoff_longitude)
        return self._haversine_distance(coords0, coords1)

    def trip_by_dow(self) -> Series:
        """
        Count trips by week day.

        :return: trip counts by day of week
        """
        trip_by_dow: Series = self.pickup_datetime.dt.dayofweek.value_counts().sort_index()
        trip_by_dow.name = "trip_by_dow"
        return trip_by_dow

    def trip_by_ih(self, step) -> Series:
        """
        Count trips every `step` hours.

        :param step: number of hours to groupby. Must be positive.
        :type step: int, float
        :return: trip counts every `step` hours
        """
        if type(step) in [int, float] and step > 0 and 24 >= step:
            trip_by_step_h: Series = self.pickup_datetime.dt.hour.apply(
                lambda hour: step * (hour // step)
            ).value_counts().sort_index()
            trip_by_step_h.name = f"trip_by_{step}h"
            return trip_by_step_h
        else:
            raise ValueError(
                f"step should int or float, and within ]0, 24]. {step} given."
            )

    def trip_by_4h(self) -> Series:
        """
        Count trips by 4h steps.

        :return: trip counts every 4hours
        """
        return self.trip_by_ih(4)

    def km_by_dow(self) -> Series:
        """
        Count kilometers traveled by week days.

        :return: km counts by day of week
        """
        copy = self.assign(day_of_week=self.pickup_datetime.dt.dayofweek)
        km_by_dow = (
            copy[["distance", "day_of_week"]].groupby("day_of_week").sum()["distance"]
        )
        km_by_dow.name = "km_by_dow"
        return km_by_dow
